### PR TITLE
Graph coloring

### PR DESCRIPTION
**Pull request by CSenshi**
### **Describe your change:**

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: backtracking/coloring.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    >>> color = 2
+    >>> valid_coloring(neighbours, colored_vertices, color)
+    False
+    """
+    for i in range(len(neighbours)):
+        # Constraint Failed
+        if neighbours[i] == 1 and colored_vertices[i] == color:
+            return False
+    # All neighbours satisfy constraint
+    return True
```
Comment (cclauss) [inline backtracking/coloring.py:36]:
```suggestion
    # Does any neighbour not satisfy the constraints
    return not any(neighbour == 1 and colored_vertices[i] == color
                   for i, neighbour in enumerate(neighbours))
```

File: backtracking/coloring.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    >>> util_color(graph, max_colors, colored_vertices, index)
+    False
+    """
+
+    # Base Case
+    if index == len(graph):
+        return True
+
+    # Recursive Step
+    for i in range(0, max_colors):
```
Comment (cclauss) [inline backtracking/coloring.py:79]:
```suggestion
    for i in range(max_colors):
```

File: backtracking/coloring.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            if util_color(graph, max_colors, colored_vertices, index + 1):
+                return True
+            # Backtrack
+            colored_vertices[index] = -1
+    return False
+
+
+def color(
+    graph: List[List[int]], max_colors: int
+) -> Tuple[bool, Union[List[int], None]]:
```
Comment (cclauss) [inline backtracking/coloring.py:93]: Isn't `typing.Optional[X]` the way to do `Union[X, None]`

Why return `None` when we could just return an empty list?

File: backtracking/coloring.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
+from typing import List, Tuple, Union
+
+"""
```
Comment (cclauss) [inline backtracking/coloring.py:3]: Please move this docstring above the imports.

File: backtracking/coloring.py (hunk 3 above)
Comment (CSenshi) [reply backtracking/coloring.py:90]: Thanks for responding! 
What do you think is better? Returning empty list of ints or leaving as it is and change to `typing.Optional[X]`

Comment (cclauss) [reply backtracking/coloring.py:90]: Empty list.  We consistently return one data type (a list).

Comment (cclauss) [reply backtracking/coloring.py:90]: Do we need to return two values?  Is there a situation where the return value is True but the list is empty or False but the list is not empty?

Some doctests on `color()` would be a welcome addition.

Comment (CSenshi) [reply backtracking/coloring.py]: There is only one case, where graph is empty list algorithm return (True, []), but it is edge case and probably no one will use it. We can change it to return one data type and when no coloring is found return empty list. Do you think it's better?

Comment (cclauss) [reply backtracking/coloring.py]: I thing it is much easier for the caller if we only return the list.

Comment (CSenshi) [reply backtracking/coloring.py]: changed :+1:

File: backtracking/coloring.py (hunk 4 above)
Comment (CSenshi) [reply backtracking/coloring.py:1]: done :+1:

**Pull request merged**